File: impl/src/env.rs
//! 环境（Environment）：变量作用域的链式结构。
//!
//! Tenet 的块（`{}`）与函数调用都会创建新的作用域，子作用域通过
//! `parent` 指针访问外层变量——这就是词法作用域（Lexical Scoping）。
//! 函数调用不捕获环境（没有闭包），因此环境可以安全地以 `Rc` 共享。

use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

use crate::error::{TenetError, TResult};
use crate::value::Value;

pub struct Env {
    vars: RefCell<HashMap<String, Value>>,
    parent: Option<Rc<Env>>,
}

impl Env {
    /// 创建一个全局环境（无父作用域）。
    pub fn global() -> Rc<Env> {
        Rc::new(Env {
            vars: RefCell::new(HashMap::new()),
            parent: None,
        })
    }

    /// 以 `parent` 为父作用域创建子环境。
    pub fn child(parent: &Rc<Env>) -> Rc<Env> {
        Rc::new(Env {
            vars: RefCell::new(HashMap::new()),
            parent: Some(Rc::clone(parent)),
        })
    }

    /// 定义变量（总是写入当前作用域，允许遮蔽外层同名变量）。
    pub fn define(&self, name: &str, value: Value) {
        self.vars.borrow_mut().insert(name.to_string(), value);
    }

    /// 读取变量，沿作用域链向上查找。
    pub fn get(&self, name: &str) -> TResult<Value> {
        if let Some(v) = self.vars.borrow().get(name) {
            return Ok(v.clone());
        }
        if let Some(parent) = &self.parent {
            return parent.get(name);
        }
        Err(TenetError::new(format!("未定义的变量 `{name}`"), None))
    }

    /// 赋值：沿作用域链找到变量并修改；找不到则报错。
    pub fn assign(&self, name: &str, value: Value) -> TResult<()> {
        if self.vars.borrow().contains_key(name) {
            self.vars.borrow_mut().insert(name.to_string(), value);
            return Ok(());
        }
        if let Some(parent) = &self.parent {
            return parent.assign(name, value);
        }
        Err(TenetError::new(format!("未定义的变量 `{name}`"), None))
    }
}

File: tenet/impl-py/tenet/env.py
"""环境（Environment）：变量作用域的链式结构。

与 Rust 版 impl-rs/src/env.rs 语义一致：

- 块（`{}`）与函数调用创建新的作用域，子作用域通过 parent 访问外层
- 定义（define）只写当前层 → 允许遮蔽
- 查找 / 赋值（get / assign）沿链向上
- 顶层不建作用域：顶层 let 直接进全局环境（REPL 跨行记忆的关键）
"""

from __future__ import annotations

from .error import TenetError


class Env:
    __slots__ = ("vars", "parent")

    def __init__(self, parent: "Env | None" = None):
        self.vars: dict[str, object] = {}
        self.parent = parent

    @classmethod
    def global_(cls) -> "Env":
        """创建一个全局环境（无父作用域）。"""
        return cls(None)

    @classmethod
    def child(cls, parent: "Env") -> "Env":
        """以 parent 为父作用域创建子环境。"""
        return cls(parent)

    def define(self, name: str, value) -> None:
        self.vars[name] = value

    def get(self, name: str) -> object:
        env = self
        while env is not None:
            if name in env.vars:
                return env.vars[name]
            env = env.parent
        raise TenetError(f"未定义的变量 `{name}`")

    def assign(self, name: str, value) -> None:
        env = self
        while env is not None:
            if name in env.vars:
                env.vars[name] = value
                return
            env = env.parent
        raise TenetError(f"未定义的变量 `{name}`")
